Runigram.java

```diff
@@ -60,48 +78,90 @@ private static void print(Color c) {
 	 */
 	public static Color[][] flippedHorizontally(Color[][] image) {
-		//// Replace the following statement with your code
-		return null;
+		int numRows = image.length;
+		int numCols = image[0].length;
+		Color[][] imageFlippedHoriz = new Color[numRows][numCols];
+		for (int row = 0; row < numRows; row++) {
+			for (int col = 0; col < numCols; col++) {
+				imageFlippedHoriz[row][col] = image[row][numCols - (col + 1)];
+			}
+		}
+		return imageFlippedHoriz;
 	}
 	
 	/**
-	 * Returns an image which is the vertically flipped version of the given image. 
+	 * Returns an image which is the vertically flipped version of the given image.
+	 * @param image
+	 * @return image vertically flipped
 	 */
 	public static Color[][] flippedVertically(Color[][] image){
-		//// Replace the following statement with your code
-		return null;
+		int numRows = image.length;
+		int numCols = image[0].length;
+		Color[][] imageFlippedVert = new Color[numRows][numCols];
+		for (int col = 0; col < numCols; col++) {
+			for (int row = 0; row < numRows; row++) {
+				imageFlippedVert[row][col] = image[numRows - (row + 1)][col];
+			}
+		}
+		return imageFlippedVert;
 	}
 	
 	// Computes the luminance of the RGB values of the given pixel, using the formula 
 	// lum = 0.299 * r + 0.587 * g + 0.114 * b, and returns a Color object consisting
 	// the three values r = lum, g = lum, b = lum.
 	public static Color luminance(Color pixel) {
-		//// Replace the following statement with your code
-		return null;
+		double lumRed = 0.299 * pixel.getRed();
+		double lumGreen = 0.587 * pixel.getGreen();
+		double lumBlue = 0.114 * pixel.getBlue();
+		int lum = (int) (lumRed + lumGreen + lumBlue);
+		Color lumPixel = new Color(lum, lum, lum);
+		return lumPixel;
 	}
 	
 	/**
 	 * Returns an image which is the grayscaled version of the given image.
 	 */
 	public static Color[][] grayScaled(Color[][] image) {
-		//// Replace the following statement with your code
-		return null;
+		int numRows = image.length;
+		int numCols = image[0].length;
+		Color[][] greyScaledImage = new Color[numRows][numCols];
+		for (int row = 0; row < numRows; row++) {
+			for (int col = 0; col < numCols; col++) {
+				greyScaledImage[row][col] = luminance(image[row][col]);
+			}
+		}
+		return greyScaledImage;
 	}	
 	
 	/**
 	 * Returns an image which is the scaled version of the given image. 
 	 * The image is scaled (resized) to have the given width and height.
 	 */
 	public static Color[][] scaled(Color[][] image, int width, int height) {
-		//// Replace the following statement with your code
-		return null;
+		int orgHeight = image.length;
+		int orgWidth = image[0].length;
+		int numRows = height;
+		int numCols = width;
+		Color[][] scaledImage = new Color[numRows][numCols];
+		for (int row = 0; row < numRows; row++) {
+			for (int col = 0; col < numCols; col++) {
+				scaledImage[row][col] = image[(int) (row * orgHeight / height)]
+											 [(int) (col * orgWidth / width)];
+			}
+		}

```

Comment:
Better to store constant calculations in variables.
```suggestion
		double widthScale = orgWidth / width;
		double heightScale = orgHeight / height;
		for (int row = 0; row < numRows; row++) {
			for (int col = 0; col < numCols; col++) {
				scaledImage[row][col] = image[(int) (row * heightScale )][(int) (col * widthScale)];
			}
		}

```